### DISCUSSION TITLE
Plotting using solely HDDs

### DISCUSSION THREAD

**Discussion by j4r3kb** [madMAx43v3r/chia-plotter#214] [Show and tell]
I have 4x Velociraptor 2.5'' 10k 1TB drives and was plotting on them with Swar Manager in the standard process in parallel 1 plot per drive and they were finishing the plot in **6.5 hours**.

I tried this plotter just for fun (wasn't expecting miracles on HDD) and here are the results for single run with 1 plot, for temp, temp2 and final separate HDD was used. 8 threads.

```
./build/chia_plot -r 8 -t /media/raptor2/ -2 /media/raptor3/ -d /media/raptor4/ -p xxx -f xxx
Multi-threaded pipelined Chia k32 plotter - 18f4918
Final Directory: /media/raptor4/
Number of Plots: 1
Process ID: 11685
Number of Threads: 8
Number of Buckets: 2^7 (128)
Pool Public Key:   xxx
Farmer Public Key: xxx
Working Directory:   /media/raptor2/
Working Directory 2: /media/raptor3/
Plot Name: plot-k32-2021-06-10-22-16-7397106a1462efb12655900c8d217bcc35cdc72371640fdcd3f494487ff37aa2
[P1] Table 1 took 183.253 sec
[P1] Table 2 took 714.764 sec, found 4295106355 matches
[P1] Table 3 took 1173.22 sec, found 4295092961 matches
[P1] Table 4 took 1385.6 sec, found 4295174532 matches
[P1] Table 5 took 1200.97 sec, found 4295168983 matches
[P1] Table 6 took 1022.64 sec, found 4295131175 matches
[P1] Table 7 took 710.215 sec, found 4295155595 matches
Phase 1 took 6391.04 sec
[P2] max_table_size = 4295174532
[P2] Table 7 scan took 234.217 sec
[P2] Table 7 rewrite took 518.844 sec, dropped 0 entries (0 %)
[P2] Table 6 scan took 127.722 sec
[P2] Table 6 rewrite took 363.486 sec, dropped 581390214 entries (13.536 %)
[P2] Table 5 scan took 137.85 sec
[P2] Table 5 rewrite took 353.2 sec, dropped 762127984 entries (17.7438 %)
[P2] Table 4 scan took 140.129 sec
[P2] Table 4 rewrite took 343.917 sec, dropped 829063661 entries (19.3022 %)
[P2] Table 3 scan took 145.449 sec
[P2] Table 3 rewrite took 353.182 sec, dropped 855175847 entries (19.9105 %)
[P2] Table 2 scan took 143.432 sec
[P2] Table 2 rewrite took 350.164 sec, dropped 865648238 entries (20.1543 %)
Phase 2 took 3219.94 sec
Wrote plot header with 268 bytes
[P3-1] Table 2 took 286.84 sec, wrote 3429458117 right entries
[P3-2] Table 2 took 367.799 sec, wrote 3429458117 left entries, 3429458117 final
[P3-1] Table 3 took 349.603 sec, wrote 3439917114 right entries
[P3-2] Table 3 took 365.751 sec, wrote 3439917114 left entries, 3439917114 final
[P3-1] Table 4 took 353.39 sec, wrote 3466110871 right entries
[P3-2] Table 4 took 367.907 sec, wrote 3466110871 left entries, 3466110871 final
[P3-1] Table 5 took 382.912 sec, wrote 3533040999 right entries
[P3-2] Table 5 took 373.093 sec, wrote 3533040999 left entries, 3533040999 final
[P3-1] Table 6 took 396.455 sec, wrote 3713740961 right entries
[P3-2] Table 6 took 422.121 sec, wrote 3713740961 left entries, 3713740961 final
[P3-1] Table 7 took 740.048 sec, wrote 4295155595 right entries
[P3-2] Table 7 took 504.198 sec, wrote 4294967296 left entries, 4294967296 final
Phase 3 took 4911.38 sec, wrote 21877235358 entries to final plot
[P4] Starting to write C1 and C3 tables
[P4] Finished writing C1 and C3 tables
[P4] Writing C2 table
[P4] Finished writing C2 table
Phase 4 took 118.035 sec, final plot size is 108835884324 bytes
Total plot creation time was 14640.4 sec
Started copy to /media/raptor4/plot-k32-2021-06-10-22-16-7397106a1462efb12655900c8d217bcc35cdc72371640fdcd3f494487ff37aa2.plot
Copy to /media/raptor4/plot-k32-2021-06-10-22-16-7397106a1462efb12655900c8d217bcc35cdc72371640fdcd3f494487ff37aa2.plot finished, took 677.453 sec, 153.212 MB/s avg.
``` 
So **4 hours**. Not worth it for my case to change the plotting process.

Now, if someone has any suggestions for another test I can do it. Maybe with 4 HDDs I could try using the same drive for temp and temp2 assign 2 threads and run 4 of the same tasks with the same destination HDD. So 8 threads total writing to common destination. What do you guys think?

**Comment by dyn4mic** [comment]
I had success with 8x146 15k SAS drives and raid0 them together, still figuring out the relationship between tmp1 and tmp2.
at the moment 3 disks tmp1 5 disks tmp2 gives me <5000s plot time. --> ~17,28 plots/day
beforehand i tried the relation ship of tmp1 to tmp2 2:6 disks gave me around ~5150s plot time --> 16,77 plots/day
until then i used Swar with 4 parallel plots mdadm 2 drives each ~7.5h per plot --> 12plots/day
currently playing around on just bundling all drives together and using just one tmp dir.
i am on a e5 2678 v3(12c/24t) with ddr3 32gb ram

**Comment by j4r3kb** [comment]
Today I have run 4 parallel plotter processes 1 plot per HDD drive, temp1 and temp2 being the same, all to common destination directory.
And finish times are 5.5 - 6 hours (including copying to destination) so up to 1 hour faster than standard process.
Now I will explore buckets less than 128.

**Comment by dyn4mic** [comment]
So my conclusion, although waste of space, in my setup a striped zfs pool of my 8 disks brings the best performance, around 4700s per plot.
Concepts are different, madmax plotter relies on fast Io and tries to saturate it, if you don't have that, same as I, and Io stays bottleneck the benefit isn't that big.
I will play around and will still compare parallel plots as well.

**Reply by j4r3kb** [reply]
In my case 1 hour is still quite nice benefit. Although I have no more space to plot ;) So just testing around.

**Reply by dyn4mic** [reply]
after some playing around. 

hardware:
xeon e5-2678 v3
32GB DDR3(4x8GB quad channel)
huananzhi x99 tf mainboard
hp h240 HBA
8x hp 146GB 15K SAS drives

using xfs with lvm(thanks to the hint of https://github.com/Chia-Network/chia-blockchain/issues/6350)

sudo vgcreate md0 /dev/sda /dev/sdb /dev/sdc /dev/sdd /dev/sde /dev/sdf /dev/sdg /dev/sdh
sudo lvcreate --type raid0 --stripes 8 --stripesize 1024 -l 100%free -y -n chiatmp md0
sudo mkfs.xfs -b size=4096 -l lazy-count=1 -f /dev/md0/chiatmp -L chiatmp
sudo mount -o sunit=256,swidth=2048 /dev/md0/chiatmp /mnt/md0

with that running at 12 threads and 128 buckets is the fastest config until now: 3014s plot time

stripesize | crc | buckets | ph1 | ph2 | ph3 | ph4 | s | min | plots/d
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
512 | 0 | 128 | 1258 | 763 | 1098 | 93 | 3212 | 53,53 | 26,90
512 | 0 | 256 | 1370 | 777 | 1209 | 99 | 3455 | 57,58 | 25,01
1024 | 1 | 64 | 1339 | 761 | 967 | 88 | 3155 | 52,58 | 27,39
1024 | 0 | 128 | 1175 | 771 | 974 | 95 | 3015 | 50,25 | 28,66
1024 | 1 | 128 | 1180 | 764 | 979 | 91 | 3014 | 50,23 | 28,67
2048 | 1 | 64 | 1343 | 783 | 938 | 90 | 3154 | 52,57 | 27,39
2048 | 1 | 128 | 1207 | 790 | 1007 | 92 | 3096 | 51,60 | 27,91
2048 | 1 | 256 | 1350 | 801 | 1178 | 100 | 3429 | 57,15 | 25,20

**Reply by vladaepro** [reply]
try -m crc=0 when formatting to xfs

**Reply by j4r3kb** [reply]
> sudo lvcreate --type raid0 --stripes 8 --stripesize 1024 -l 100%free -y -n chiatmp md0

is this line correct? around `-l 100%free -y`?

**Reply by dyn4mic** [reply]
yes -y parameter to skip a prompt of accepting overwrite of other fs signature

**Reply by j4r3kb** [reply]
I mean `100%free` strange value ;)

**Reply by dyn4mic** [reply]
yeah that one as well, allocating 100% of free space full in line with spec https://linux.die.net/man/8/lvcreate

**Reply by j4r3kb** [reply]
Using your tips on xfs and raid0 I have created volume from 3 x WD1000DHTZ and 1 x ST2000NC001.
Run 1 plot with 6 threads, 128 buckets.
```
Multi-threaded pipelined Chia k32 plotter - f495816
Final Directory: /media/farm/plot/
Number of Plots: 1
Crafting plot 1 out of 1
Process ID: 8942
Number of Threads: 6
Number of Buckets: 2^7 (128)
Pool Public Key:   xxx
Farmer Public Key: xxx
Working Directory:   /media/chiatmp/
Working Directory 2: /media/chiatmp/
Plot Name: plot-k32-2021-06-16-23-54-345c5d3cb99ae9f5cb7d377d4f6efe0b018f415069f3ba6f8249fb6fb0bfa7df
[P1] Table 1 took 59.5048 sec
[P1] Table 2 took 300.973 sec, found 4294928466 matches
[P1] Table 3 took 458.932 sec, found 4294971347 matches
[P1] Table 4 took 546.559 sec, found 4294848206 matches
[P1] Table 5 took 494.747 sec, found 4294667963 matches
[P1] Table 6 took 429.475 sec, found 4294342772 matches
[P1] Table 7 took 290.852 sec, found 4293791925 matches
Phase 1 took 2581.87 sec
[P2] max_table_size = 4294971347
[P2] Table 7 scan took 80.2514 sec
[P2] Table 7 rewrite took 168.154 sec, dropped 0 entries (0 %)
[P2] Table 6 scan took 54.9719 sec
[P2] Table 6 rewrite took 148.112 sec, dropped 581335561 entries (13.5372 %)
[P2] Table 5 scan took 49.2346 sec
[P2] Table 5 rewrite took 133.652 sec, dropped 762012975 entries (17.7432 %)
[P2] Table 4 scan took 64.1322 sec
[P2] Table 4 rewrite took 132.591 sec, dropped 828886357 entries (19.2995 %)
[P2] Table 3 scan took 48.3343 sec
[P2] Table 3 rewrite took 139.585 sec, dropped 855099123 entries (19.9093 %)
[P2] Table 2 scan took 47.7997 sec
[P2] Table 2 rewrite took 142.889 sec, dropped 865532370 entries (20.1524 %)
Phase 2 took 1214.42 sec
Wrote plot header with 268 bytes
[P3-1] Table 2 took 176.591 sec, wrote 3429396096 right entries
[P3-2] Table 2 took 159.621 sec, wrote 3429396096 left entries, 3429396096 final
[P3-1] Table 3 took 205.689 sec, wrote 3439872224 right entries
[P3-2] Table 3 took 162.211 sec, wrote 3439872224 left entries, 3439872224 final
[P3-1] Table 4 took 201.106 sec, wrote 3465961849 right entries
[P3-2] Table 4 took 159.102 sec, wrote 3465961849 left entries, 3465961849 final
[P3-1] Table 5 took 204.055 sec, wrote 3532654988 right entries
[P3-2] Table 5 took 152.147 sec, wrote 3532654988 left entries, 3532654988 final
[P3-1] Table 6 took 216.602 sec, wrote 3713007211 right entries
[P3-2] Table 6 took 168.003 sec, wrote 3713007211 left entries, 3713007211 final
[P3-1] Table 7 took 250.907 sec, wrote 4293791925 right entries
[P3-2] Table 7 took 220.352 sec, wrote 4293791925 left entries, 4293791925 final
Phase 3 took 2282.69 sec, wrote 21874684293 entries to final plot
[P4] Starting to write C1 and C3 tables
[P4] Finished writing C1 and C3 tables
[P4] Writing C2 table
[P4] Finished writing C2 table
Phase 4 took 91.4301 sec, final plot size is 108820310880 bytes
Total plot creation time was 6170.45 sec (102.841 min)
Started copy to /media/farm/plot/plot-k32-2021-06-16-23-54-345c5d3cb99ae9f5cb7d377d4f6efe0b018f415069f3ba6f8249fb6fb0bfa7df.plot
Copy to /media/farm/plot/plot-k32-2021-06-16-23-54-345c5d3cb99ae9f5cb7d377d4f6efe0b018f415069f3ba6f8249fb6fb0bfa7df.plot finished, took 321.711 sec, 322.585 MB/s avg.
```

So around 14 plots per day, running parallel with standard method 1 per drive results in more plots.

**Comment by gryan315** [comment]
I ran a test with 16x 10k RPM drives in a raid0 array, saw around 2k r/w speeds while it was plotting. Array made with mdadm, formatted in xfs. The drives are 10x 600gb 64mb cache, and 6x 900gb 128mb cache, Just used a 100G partition from each in the array. 2x xeon 2678v3.

48 threads
128 buckets
Phase 1: 717.045 sec
Phase 2: 352.675 sec
Phase 3: 518.501 sec
Phase 4: 100.896 sec
Total: 1689.21 sec

For reference, using RAM tmpfs with 48 threads I only got 1566.97 sec total time.

**Reply by cojarbi** [reply]
@gryan315 I’m very disappointed to see your performance, I’m getting almost the same result with half the threads Xeon cpu. (It’s a 2.5ghz not sure the model atm but it’s old ). Reason I say this is I just ordered a 20 core thread machine hoping more thread would make it much faster

Adding an nvme did not made any improvements.

**Reply by gryan315** [reply]
We seem to be limited by IO speed rather than threads, however more threads should be able to handle more IO, so adding more drives to plot on should improve performance (IE splitting up 2 large arrays on 2 high thread numa nodes will likely yield the best results).

**Reply by cojarbi** [reply]
Actually I read this very wrong. You are doing half my time with my current setup which is only 24 threads. How exactly are you creating that raid in madam, can you share the creating command / filesystem including the cache please

**Reply by gryan315** [reply]
The cache I mentioned is the drives' built in cache, the 128mb models are newer SAS 12GB/s drives and are noticeably faster than the older drives with 64mb cache, even running at SAS 6GB/s. First I created 100gb partitions with fdisk on each of the drives using the +100G for the Last sector option. Once all the drives had a 100G partition, I used mdadm to create the array: 

`sudo mdadm --create --verbose --level=0 --raid-devices=16 /dev/md11 /dev/sdm1 /dev/sdn1 /dev/sdo1 /dev/sdp1 /dev/sdq1 /dev/sdr1 /dev/sds1 /dev/sdt1 /dev/sdu1 /dev/sdv1 /dev/sdw1 /dev/sdx1 /dev/sdy1 /dev/sdz1 /dev/sdaa1 /dev/sdab1`

Once the array was running, formatted in xfs with: `sudo mkfs.xfs /dev/md11` although it has been brought to my attention that disabling checksum metadata may improve performance, that would be `sudo mkfs.xfs -m crc=0 /dev/md11`

I may make another attempt with 2x 8 drive arrays plotting in parallel, split by numactl to each socket. Plotting 2 parallel with the 16x drive array resulted in 56 minutes each, and using a 2x nvme raid array as tmp1 reduced that to 44 minutes each.
Edit: actually now that I'm looking at it, 56 minutes each with 24 threads is perfect scaling with 28 minutes for 48 threads... Probably a waste of time to test 2x 8 drive arrays.

**Reply by cojarbi** [reply]
This makes sense, can you share how did you make those partitions? I would like to replicate as much as possible to compre. Thanks

**Reply by gryan315** [reply]
`fdisk /dev/device` then 'g' to create a new gpt partition table, and overwrite any existing table, then 'n' to create a new partition, use the defaults for partition number and First sector by hitting `enter` twice, then input +100G for Last sector. If you already had a filesystem on the drive, you will be asked if you want to remove the signature, answer 'y', then type 'w' to write the new partition table. Then repeat for the next drive. Always triple check that you are on the correct device, as fdisk can be unforgiving.

**Reply by dyn4mic** [reply]
@gryan315 arent you running to many threads? it should only cover "real" cores, plotter spins up the necessary threads on his own, threads value is just a multiplier. at least what i have seen to many threads decreases speed.
e5-2678v3 has 12c/24t each, so i expect you should do max 24threads as parameter for the plotter

**Reply by gryan315** [reply]
I just did some quick testing with the -r flag using ram tmpfs and nvme with -r 24 vs -r 48.
-r 24:
Phase 1: 680s
Phase 2: 387s
Phase 3: 349s
Phase 4: 76s
Total: 1493s

-r 48
Phase 1: 679s
Phase 2: 298s
Phase 3: 517s
Phase 4: 71s
Total: 1566s

So as you can see, phase 1 and 4 are unaffected, phase 2 is faster with logical cores, and phase 3 is faster with physical cores. Phase 3's speed up is more than phase 2's slow down, so slight improvement overall with physical cores as the -r value. I guess optimal would be if you could set phase 2 to use double the multiplier of the rest of the job.

Edit: I also just ran a test using tmpfs and nvme in parallel, comparable to my test of tmpfs and 16x array in parallel, and the results were worse: using -r 12 for both jobs, phase 2 and phase 3 were both slower than when I used -r 24 for both jobs previously. Phase 1 and 4 were roughly the same for both. Total slowdown of about 2 minutes each.

**Reply by cojarbi** [reply]
How about using the -u and -v options differently, would that make a difference?

**Reply by gryan315** [reply]
I didn't see a ton of difference in performance changing buckets. I need to run at least 256 when running the tmpfs since I only have 128gb ram. I run 512 on the plot with the tmpfs because that was slightly faster than 256. Based on my testing with the NVME drives and buckets, reducing the buckets make phase 2 slightly faster (30 seconds per drop in buckets) and made phase 3 take longer.

**Reply by cojarbi** [reply]
What im saying is that with this options you can do for example -u 256 -v 512

**Reply by gryan315** [reply]
Yes, I'm aware what the flags do, but in my testing, it doesn't make a significant difference. If anything I would want phase 2 and 3 to be 128 buckets for a very slight improvement.

**Reply by gryan315** [reply]
I'll also add that I've split my 16x array into 10x 600gb with 64mb cache and 6x 900gb with 128mb cache, and the 6x array is slightly slower than the 10x array when used as tmp1 with ramdisk for -2.

**Reply by madMAx43v3r** [reply]
> So as you can see, phase 1 and 4 are unaffected, phase 2 is faster with logical cores, and phase 3 is faster with physical cores. Phase 3's speed up is more than phase 2's slow down, so slight improvement overall with physical cores as the -r value. I guess optimal would be if you could set phase 2 to use double the multiplier of the rest of the job.

See https://github.com/madMAx43v3r/chia-plotter/issues/611

**Reply by gryan315** [reply]
Sounds exciting

**Reply by cojarbi** [reply]
I see - k was added, how what are the parameter options for it?

**Reply by gryan315** [reply]
> I see - k was added, how what are the parameter options for it?

I haven't had a chance to test it yet, but it would appear to me that you want to set it to 2 to double phase 2 -r value.

**Reply by gryan315** [reply]
Just finished the first plot using -K, it works as expected. Using -r 24 and -k 2 resulted in phase 2 times the same as -r 48, while maintaining the faster phase 3 time with -r 24. New fastest plot for my rig.

**Comment by cojarbi** [comment]
So i got my new plotter a Dell R Dual Xeon E5-2670 v2 with 128GB ram and SAS 15k HDD. Using tmpfs and hardware raid0 on a h710 mini.

My first set of testing brings speed improvements over my previous one with this results:

256 -r 40
787/546/639/71 > 2043

256+512 -r 40
796/563/700/58 > 2117

512 -r 40
823/644/715/57> 2239

512+256 -r 40
846/590/691/68> 2195

Now i started testing -r 20 and its the same average result

**Reply by gryan315** [reply]
You may want to give -r 30 a shot, I've had decent results with -r 36 being faster than both -r 24 and -r 48. Keep experimenting if you've got the time.

**Reply by cojarbi** [reply]
I se a difference of about 2 min less when using 36 and about the same as before when using 30. I need to try clearing my hardware raid and using mdadm

**Reply by gryan315** [reply]
Can you do that with an h710, thought it was raid only? Are you planning to make a bunch of single drive raid 0 arrays on the h710 and then mdadm them in linux? If possible, I'd recommend swapping out the h710 for an h310 mini, or if you have a full size pcie an h200, which is what I use (the h200e external version). I'm not sure if there is a significant performance difference between hardware raid and software raid on a HBA, at least not when using the older hardware it was designed to be used with.

**Reply by cojarbi** [reply]
i have another server with a h310, not much difference. Yes you can flash most of those cards, this works and tested and its reversible, simple boot and flash https://fohdeesha.com/docs/perc/

The reason i say this is that i was able to see max w/r speed with mdadm in the 2000 range and with the cards is at 600 maybe

**Reply by cojarbi** [reply]
I haven't been able to test much, the server freezes when im sudo mkfs.xfs -m crc=0 /dev/md11

**Reply by gryan315** [reply]
You can try without the -m crc=0, in my quick testing I saw no significant advantage to using it.
